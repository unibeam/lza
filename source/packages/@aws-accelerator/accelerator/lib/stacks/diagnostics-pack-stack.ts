import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NagSuppressions } from 'cdk-nag';
import path from 'path';

export interface DiagnosticsPackStackProps extends cdk.StackProps {
  readonly acceleratorPrefix: string;
  readonly ssmParamPrefix: string;
  readonly bucketNamePrefix: string;
  readonly installerStackName: string;
  readonly configRepositoryName: string;
  readonly qualifier?: string;
}

export class DiagnosticsPackStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: DiagnosticsPackStackProps) {
    super(scope, id, props);

    let keyParameterName = `${props.ssmParamPrefix}/installer/kms/key-arn`;
    let reportDestinationBucketName = `${props.bucketNamePrefix}-installer-${cdk.Stack.of(this).account}-${
      cdk.Stack.of(this).region
    }`;
    let pipelineAccountResourcesPrefix = props.acceleratorPrefix;
    let diagnosticsPackLambdaRoleNamePrefix = props.acceleratorPrefix;

    if (props.qualifier) {
      keyParameterName = `${props.ssmParamPrefix}/${props.qualifier}/installer/kms/key-arn`;
      reportDestinationBucketName = `${props.qualifier}-installer-${cdk.Stack.of(this).account}-${
        cdk.Stack.of(this).region
      }`;
      pipelineAccountResourcesPrefix = props.qualifier;
      diagnosticsPackLambdaRoleNamePrefix = props.qualifier;
    }

    const destinationBucket = cdk.aws_s3.Bucket.fromBucketName(this, 'DestinationBucket', reportDestinationBucketName);
    const diagnosticsPackAccessRole = `${props.acceleratorPrefix}-DiagnosticsPackAccessRole`;

    const diagnosticsPackLambdaRole = new cdk.aws_iam.Role(this, 'DiagnosticsPackLambdaRole', {
      roleName: `${diagnosticsPackLambdaRoleNamePrefix}-DiagnosticsPackLambdaRole`,
      assumedBy: new cdk.aws_iam.ServicePrincipal(`lambda.amazonaws.com`),
      managedPolicies: [cdk.aws_iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')],
    });

    diagnosticsPackLambdaRole.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['sts:AssumeRole'],
        resources: [`arn:${cdk.Stack.of(this).partition}:iam::*:role/${diagnosticsPackAccessRole}`],
      }),
    );

    diagnosticsPackLambdaRole.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'AllowStsCallerIdentityActions',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['sts:GetCallerIdentity'],
        resources: ['*'],
      }),
    );

    diagnosticsPackLambdaRole.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'CloudformationAccess',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['cloudformation:DescribeStackEvents', 'cloudformation:DescribeStacks'],
        resources: [
          `arn:${cdk.Stack.of(this).partition}:cloudformation:${cdk.Stack.of(this).region}:${
            cdk.Stack.of(this).account
          }:stack/${pipelineAccountResourcesPrefix}*`,
        ],
      }),
    );

    if (!props.qualifier) {
      diagnosticsPackLambdaRole.addToPrincipalPolicy(
        new cdk.aws_iam.PolicyStatement({
          sid: 'OrganizationsAccess',
          effect: cdk.aws_iam.Effect.ALLOW,
          actions: ['organizations:ListAccounts'],
          resources: ['*'],
        }),
      );
    }

    diagnosticsPackLambdaRole.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'CodecommitAccess',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['codecommit:GetFile'],
        resources: [
          `arn:${cdk.Stack.of(this).partition}:codecommit:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:${
            props.configRepositoryName
          }`,
        ],
      }),
    );

    diagnosticsPackLambdaRole.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'CodepipelineAccess',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['codepipeline:GetPipelineState'],
        resources: [
          `arn:${cdk.Stack.of(this).partition}:codepipeline:${cdk.Stack.of(this).region}:${
            cdk.Stack.of(this).account
          }:${pipelineAccountResourcesPrefix}*`,
        ],
      }),
    );

    diagnosticsPackLambdaRole.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'CloudwatchLogsAccess',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['logs:FilterLogEvents'],
        resources: [
          `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
            cdk.Stack.of(this).account
          }:log-group:/aws/codebuild/${pipelineAccountResourcesPrefix}*`,
          `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
            cdk.Stack.of(this).account
          }:log-group:/aws/lambda/${pipelineAccountResourcesPrefix}*`,
        ],
      }),
    );

    diagnosticsPackLambdaRole.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'DiagnosticsBucketWriteAccess',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: [
          's3:PutObject',
          's3:GetObject',
          's3:AbortMultipartUpload',
          's3:ListBucket',
          's3:DeleteObject',
          's3:GetObjectVersion',
          's3:ListMultipartUploadParts',
        ],
        resources: [destinationBucket.bucketArn, destinationBucket.arnForObjects('*')],
      }),
    );

    // AwsSolutions-IAM4: The IAM user, role, or group uses AWS managed policies.
    NagSuppressions.addResourceSuppressionsByPath(this, `${this.stackName}/DiagnosticsPackLambdaRole/Resource`, [
      {
        id: 'AwsSolutions-IAM4',
        reason: 'Custom resource Lambda role policy.',
      },
    ]);

    // AwsSolutions-IAM5: The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.
    NagSuppressions.addResourceSuppressionsByPath(
      this,
      `${this.stackName}/DiagnosticsPackLambdaRole/DefaultPolicy/Resource`,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Custom resource Lambda role policy.',
        },
      ],
    );

    const diagnosticFunction = new cdk.aws_lambda.Function(this, 'DiagnosticsFunction', {
      role: diagnosticsPackLambdaRole,
      description: 'Accelerator diagnostics report lambda function.',
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, '../lambdas/diagnostic-pack/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      memorySize: 512,
      timeout: cdk.Duration.minutes(15),
      handler: 'index.handler',
      environment: {
        INSTALLER_STACK_NAME: props.installerStackName,
        HOME_REGION: cdk.Stack.of(this).region,
        PIPELINE_ACCOUNT_ID: cdk.Stack.of(this).account,
        PARTITION: cdk.Stack.of(this).partition,
        DAYS_PIPELINE_IN_FAILED_STATUS: '1',
        REPORT_BUCKET_NAME: destinationBucket.bucketName,
        CONFIG_REPO_NAME: props.configRepositoryName,
        MANAGEMENT_ACCOUNT_ROLE_NAME: diagnosticsPackAccessRole,
      },
    });

    new cdk.aws_logs.LogGroup(this, `${diagnosticFunction.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${diagnosticFunction.functionName}`,
      retention: 30,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const key = cdk.aws_kms.Key.fromKeyArn(
      this,
      'DiagnosticsProjectEncryptionKey',
      cdk.aws_ssm.StringParameter.valueForStringParameter(this, keyParameterName).toString(),
    );

    const buildProject = new cdk.aws_codebuild.Project(this, 'DiagnosticsProject', {
      projectName: `${pipelineAccountResourcesPrefix}-DiagnosticProject`,
      encryptionKey: key,
      description: `Accelerator diagnostic project. You can execute this project to generate an error report and store it into the ${destinationBucket.bucketName} bucket.`,
      buildSpec: cdk.aws_codebuild.BuildSpec.fromObjectToYaml({
        version: '0.2',
        phases: {
          build: {
            commands: [
              `set -e`,
              `aws lambda update-function-configuration --function-name  ${diagnosticFunction.functionArn}  --region ${
                cdk.Stack.of(this).region
              } --environment Variables="{INSTALLER_STACK_NAME=${props.installerStackName},HOME_REGION=${
                cdk.Stack.of(this).region
              },PIPELINE_ACCOUNT_ID=${cdk.Stack.of(this).account},PARTITION=${
                cdk.Stack.of(this).partition
              },DAYS_PIPELINE_IN_FAILED_STATUS=$DAYS_PIPELINE_IN_FAILED_STATUS,REPORT_BUCKET_NAME=${
                destinationBucket.bucketName
              },CONFIG_REPO_NAME=${
                props.configRepositoryName
              },MANAGEMENT_ACCOUNT_ROLE_NAME=${diagnosticsPackAccessRole}}" --output text`,
              `aws lambda wait function-updated --function-name ${diagnosticFunction.functionArn} --region ${
                cdk.Stack.of(this).region
              }`,
              `aws lambda invoke --function-name ${diagnosticFunction.functionArn} --region ${
                cdk.Stack.of(this).region
              } --payload {} /tmp/response.json`,
              `error_count=$(grep "error"  /tmp/response.json | wc -l)`,
              `echo $error_count`,
              `if [ $error_count -gt 0 ]; then 
                echo "Diagnostics Lambda execution failed with below error !!!!"; 
                cat /tmp/response.json; 
                exit 1; 
                fi`,
            ],
          },
        },
      }),
      environment: {
        buildImage: cdk.aws_codebuild.LinuxBuildImage.STANDARD_7_0,
        privileged: false,
        computeType: cdk.aws_codebuild.ComputeType.SMALL,
        environmentVariables: {
          DAYS_PIPELINE_IN_FAILED_STATUS: {
            type: cdk.aws_codebuild.BuildEnvironmentVariableType.PLAINTEXT,
            value: '1',
          },
        },
      },
    });

    buildProject.role?.addToPrincipalPolicy(
      new cdk.aws_iam.PolicyStatement({
        sid: 'AllowLambdaAccess',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['lambda:InvokeFunction', 'lambda:UpdateFunctionConfiguration', 'lambda:GetFunctionConfiguration'],
        resources: [diagnosticFunction.functionArn],
      }),
    );

    // AwsSolutions-IAM5: The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.
    NagSuppressions.addResourceSuppressionsByPath(
      this,
      `${this.stackName}/DiagnosticsProject/Role/DefaultPolicy/Resource`,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Diagnostic CodeBuild project role.',
        },
      ],
    );
  }
}
