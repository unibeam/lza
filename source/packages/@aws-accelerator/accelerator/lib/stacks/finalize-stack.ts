/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { AcceleratorStack, AcceleratorStackProps, AcceleratorKeyType } from './accelerator-stack';
import { DetachQuarantineScp } from '../detach-quarantine-scp';
import { ScpResource } from '../resources/scp-resource';

export class FinalizeStack extends AcceleratorStack {
  constructor(scope: Construct, id: string, props: AcceleratorStackProps) {
    super(scope, id, props);

    if (props.globalRegion === cdk.Stack.of(this).region) {
      this.logger.debug(`Retrieving CloudWatch kms key`);

      const lambdaKey = this.getAcceleratorKey(AcceleratorKeyType.LAMBDA_KEY);
      const cloudwatchKey = this.getAcceleratorKey(AcceleratorKeyType.CLOUDWATCH_KEY);
      const scpResource = new ScpResource(this, cloudwatchKey, lambdaKey, props);

      //
      // Update SCP with dynamic parameters
      //
      scpResource.createAndAttachScps(props);

      //
      // Configure revert scp changes rule
      //
      scpResource.configureRevertScpChanges(props);

      if (process.env['CONFIG_COMMIT_ID']) {
        this.logger.debug(`Storing configuration commit id in SSM`);
        new cdk.aws_ssm.StringParameter(this, 'AcceleratorCommitIdParameter', {
          parameterName: `${props.prefixes.ssmParamName}/configuration/configCommitId`,
          stringValue: process.env['CONFIG_COMMIT_ID'],
          description: `The hash of the latest ${props.configRepositoryName} version of the LZA configuration files to deploy successfully. If you use S3 as a location for LZA config files, this will be an S3 version Id.`,
        });
      }

      if (props.organizationConfig.quarantineNewAccounts?.enable && props.partition === 'aws') {
        this.logger.debug(`Creating resources to detach quarantine scp`);
        const policyId = cdk.aws_ssm.StringParameter.valueForStringParameter(
          this,
          `${props.prefixes.ssmParamName}/organizations/scp/${props.organizationConfig.quarantineNewAccounts?.scpPolicyName}/id`,
        );

        new DetachQuarantineScp(this, 'DetachQuarantineScp', {
          scpPolicyId: policyId,
          managementAccountId: props.accountsConfig.getManagementAccountId(),
          partition: props.partition,
          kmsKey: cloudwatchKey,
          logRetentionInDays: props.globalConfig.cloudwatchLogRetentionInDays,
        });
      }

      //
      // Create NagSuppressions
      //
      this.addResourceSuppressionsByPath();
    }
    this.logger.info('Completed stack synthesis');
  }
}
