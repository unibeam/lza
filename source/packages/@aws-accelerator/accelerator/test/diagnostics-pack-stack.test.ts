/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';

import { DiagnosticsPackStack } from '../lib/stacks/diagnostics-pack-stack';
import { snapShotTest } from './snapshot-test';
import { describe } from '@jest/globals';

const testNamePrefix = 'Construct(DiagnosticsPackStack): ';

/**
 * DiagnosticsPack Stack
 */
const getStack = () => {
  const app = new cdk.App();
  const stack = new DiagnosticsPackStack(app, 'DiagnosticsPackStack', {
    acceleratorPrefix: 'AWSAccelerator',
    ssmParamPrefix: '/accelerator',
    bucketNamePrefix: 'aws-accelerator',
    installerStackName: 'AWSAccelerator-InstallerStack',
    configRepositoryName: 'aws-accelerator-config',
    qualifier: 'aws-accelerator',
    env: {
      account: '000000000000',
      region: 'us-east-1',
    },
  });
  return stack;
};
describe('DiagnosticsPackStack', () => {
  snapShotTest(testNamePrefix, getStack);
});
