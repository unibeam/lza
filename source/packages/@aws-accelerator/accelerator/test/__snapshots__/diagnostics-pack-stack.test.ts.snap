// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiagnosticsPackStack Construct(DiagnosticsPackStack):  Snapshot Test 1`] = `
{
  "Parameters": {
    "SsmParameterValueacceleratorawsacceleratorinstallerkmskeyarnC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/accelerator/aws-accelerator/installer/kms/key-arn",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DiagnosticsFunctionF6482E72": {
      "DependsOn": [
        "DiagnosticsPackLambdaRoleDefaultPolicy93B3F548",
        "DiagnosticsPackLambdaRole5CC68755",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-000000000000-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Accelerator diagnostics report lambda function.",
        "Environment": {
          "Variables": {
            "CONFIG_REPO_NAME": "aws-accelerator-config",
            "DAYS_PIPELINE_IN_FAILED_STATUS": "1",
            "HOME_REGION": "us-east-1",
            "INSTALLER_STACK_NAME": "AWSAccelerator-InstallerStack",
            "MANAGEMENT_ACCOUNT_ROLE_NAME": "AWSAccelerator-DiagnosticsPackAccessRole",
            "PARTITION": {
              "Ref": "AWS::Partition",
            },
            "PIPELINE_ACCOUNT_ID": "000000000000",
            "REPORT_BUCKET_NAME": "aws-accelerator-installer-000000000000-us-east-1",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "DiagnosticsPackLambdaRole5CC68755",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DiagnosticsFunctionLogGroup65B6B182": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "DiagnosticsFunctionF6482E72",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "DiagnosticsPackLambdaRole5CC68755": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "Custom resource Lambda role policy.",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "aws-accelerator-DiagnosticsPackLambdaRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "DiagnosticsPackLambdaRoleDefaultPolicy93B3F548": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Custom resource Lambda role policy.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iam::*:role/AWSAccelerator-DiagnosticsPackAccessRole",
                  ],
                ],
              },
            },
            {
              "Action": "sts:GetCallerIdentity",
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "AllowStsCallerIdentityActions",
            },
            {
              "Action": [
                "cloudformation:DescribeStackEvents",
                "cloudformation:DescribeStacks",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":cloudformation:us-east-1:000000000000:stack/aws-accelerator*",
                  ],
                ],
              },
              "Sid": "CloudformationAccess",
            },
            {
              "Action": "codecommit:GetFile",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codecommit:us-east-1:000000000000:aws-accelerator-config",
                  ],
                ],
              },
              "Sid": "CodecommitAccess",
            },
            {
              "Action": "codepipeline:GetPipelineState",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codepipeline:us-east-1:000000000000:aws-accelerator*",
                  ],
                ],
              },
              "Sid": "CodepipelineAccess",
            },
            {
              "Action": "logs:FilterLogEvents",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:000000000000:log-group:/aws/codebuild/aws-accelerator*",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:000000000000:log-group:/aws/lambda/aws-accelerator*",
                    ],
                  ],
                },
              ],
              "Sid": "CloudwatchLogsAccess",
            },
            {
              "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:AbortMultipartUpload",
                "s3:ListBucket",
                "s3:DeleteObject",
                "s3:GetObjectVersion",
                "s3:ListMultipartUploadParts",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::aws-accelerator-installer-000000000000-us-east-1",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::aws-accelerator-installer-000000000000-us-east-1/*",
                    ],
                  ],
                },
              ],
              "Sid": "DiagnosticsBucketWriteAccess",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DiagnosticsPackLambdaRoleDefaultPolicy93B3F548",
        "Roles": [
          {
            "Ref": "DiagnosticsPackLambdaRole5CC68755",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DiagnosticsProject5D286B2E": {
      "Properties": {
        "Artifacts": {
          "Type": "NO_ARTIFACTS",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "Description": "Accelerator diagnostic project. You can execute this project to generate an error report and store it into the aws-accelerator-installer-000000000000-us-east-1 bucket.",
        "EncryptionKey": {
          "Ref": "SsmParameterValueacceleratorawsacceleratorinstallerkmskeyarnC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "EnvironmentVariables": [
            {
              "Name": "DAYS_PIPELINE_IN_FAILED_STATUS",
              "Type": "PLAINTEXT",
              "Value": "1",
            },
          ],
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "aws-accelerator-DiagnosticProject",
        "ServiceRole": {
          "Fn::GetAtt": [
            "DiagnosticsProjectRole29C23DD8",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": {
            "Fn::Join": [
              "",
              [
                "version: "0.2"
phases:
  build:
    commands:
      - set -e
      - aws lambda update-function-configuration --function-name  ",
                {
                  "Fn::GetAtt": [
                    "DiagnosticsFunctionF6482E72",
                    "Arn",
                  ],
                },
                "  --region us-east-1 --environment Variables="{INSTALLER_STACK_NAME=AWSAccelerator-InstallerStack,HOME_REGION=us-east-1,PIPELINE_ACCOUNT_ID=000000000000,PARTITION=",
                {
                  "Ref": "AWS::Partition",
                },
                ",DAYS_PIPELINE_IN_FAILED_STATUS=$DAYS_PIPELINE_IN_FAILED_STATUS,REPORT_BUCKET_NAME=aws-accelerator-installer-000000000000-us-east-1,CONFIG_REPO_NAME=aws-accelerator-config,MANAGEMENT_ACCOUNT_ROLE_NAME=AWSAccelerator-DiagnosticsPackAccessRole}" --output text
      - aws lambda wait function-updated --function-name ",
                {
                  "Fn::GetAtt": [
                    "DiagnosticsFunctionF6482E72",
                    "Arn",
                  ],
                },
                " --region us-east-1
      - aws lambda invoke --function-name ",
                {
                  "Fn::GetAtt": [
                    "DiagnosticsFunctionF6482E72",
                    "Arn",
                  ],
                },
                "REPLACED-JSON-PATH.json",
              ],
            ],
          },
          "Type": "NO_SOURCE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "DiagnosticsProjectRole29C23DD8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "DiagnosticsProjectRoleDefaultPolicyEA9CD8D1": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Diagnostic CodeBuild project role.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:000000000000:log-group:/aws/codebuild/",
                      {
                        "Ref": "DiagnosticsProject5D286B2E",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:000000000000:log-group:/aws/codebuild/",
                      {
                        "Ref": "DiagnosticsProject5D286B2E",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:000000000000:report-group/",
                    {
                      "Ref": "DiagnosticsProject5D286B2E",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "SsmParameterValueacceleratorawsacceleratorinstallerkmskeyarnC96584B6F00A464EAD1953AFF4B05118Parameter",
              },
            },
            {
              "Action": [
                "lambda:InvokeFunction",
                "lambda:UpdateFunctionConfiguration",
                "lambda:GetFunctionConfiguration",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "DiagnosticsFunctionF6482E72",
                  "Arn",
                ],
              },
              "Sid": "AllowLambdaAccess",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DiagnosticsProjectRoleDefaultPolicyEA9CD8D1",
        "Roles": [
          {
            "Ref": "DiagnosticsProjectRole29C23DD8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
